lib/tpm_policy.c

```diff
@@ -0,0 +1,254 @@
+        }
+        dgst_cnt++;
+
+        total_indices_for_this_alg=0;
+    }
+
+    int file_size_test = (pcr_values->count == 0) ? 1:0;
+    if (file_size_test && pctx->pcr_policy_options.is_raw_pcrs_file) {
+        long filesize = 0;
+        bool result = files_get_file_size(pctx->pcr_policy_options.raw_pcrs_file, &filesize);
+        if (!result) {
+                return false;
+        } else {
+            if (filesize != expected_pcr_input_file_size) {
+                LOG_ERR("pcr-input-file filesize does not match pcr set-list");
+                return false;
+            } 
+        }
+    } 
+    return true;
+}
+
+TPM_RC build_pcr_policy(CREATE_POLICY_CTX *pctx) {
+    //PCR inputs validation
+    if (pctx->pcr_policy_options.is_set_list == false) {
+        LOG_ERR("Need the pcr list to account for in the policy.");
+        return TPM_RC_NO_RESULT;
+    }
+
+    // Calculate digest( with authhash alg) of pcrvalues in variable pcr_digest
+    TPM_RC rval=0;
+    TPM2B_DIGEST pcr_digest = TPM2B_TYPE_INIT(TPM2B_DIGEST, buffer);
+    TPML_DIGEST pcr_values;
+
+    bool result = evaluate_populate_pcr_digests(pctx, &pcr_values);
+    if (!result) {
+        return TPM_RC_NO_RESULT;
+    } 
+
+    if (pctx->pcr_policy_options.is_raw_pcrs_file) {
+        FILE *fp = fopen (pctx->pcr_policy_options.raw_pcrs_file, "rb");
+        if (fp == NULL) {
+            LOG_ERR("Cannot open pcr-input-file %s", pctx->pcr_policy_options.raw_pcrs_file);
+            return TPM_RC_NO_RESULT;
+        }
+       // Bank hashAlg values dictates the order of the list of digests
+        unsigned i;
+        for(i=0; i<pcr_values.count; i++) {
+            size_t sz = fread(&pcr_values.digests[i].t.buffer, 1, pcr_values.digests[i].t.size, fp);
+            if (sz != pcr_values.digests[i].t.size) {
+                const char *msg = ferror(fp) ? strerror(errno) :
+                        "end of file reached";
+                LOG_ERR("Reading from file \"%s\" failed: %s",
+                        pctx->pcr_policy_options.raw_pcrs_file, msg);
+                fclose(fp);
+                return TPM_RC_NO_RESULT;
+            }
+        }
+        fclose(fp);
+    } else {
+        UINT32 pcr_update_counter;
+        TPML_PCR_SELECTION pcr_selection_out;
+        // Read PCRs
+        TPM_RC rval = Tss2_Sys_PCR_Read(pctx->sapi_context, 0, 
+            &pctx->pcr_policy_options.pcr_selections, 
+            &pcr_update_counter, &pcr_selection_out, &pcr_values, 0);
+        if (rval != TPM_RC_SUCCESS) {
+            return rval;
+        }
+    }
+    // Calculate hashes
+    rval = tpm_hash_sequence(pctx->sapi_context, 
+        pctx->common_policy_options.policy_session->authHash, pcr_values.count, 
+        &pcr_values.digests[0], &pcr_digest);
+    if (rval != TPM_RC_SUCCESS) {
+        return rval;
+    }
+    // Call the PolicyPCR command
+    rval = Tss2_Sys_PolicyPCR(pctx->sapi_context,
```

Comment:
this can be a tail call... which can be optimized by the compiler.

return Tss2_Sys_PolicyPCR();